### PR TITLE
MRG ENH: Center and allow ch_type=None

### LINKED ISSUE #2027: ENH ch_type = None should default to channel types presented
when using ica.plot_components, the plotting function should not default to meg especially when there are no meg channels present in the data. the default ch_type should be None. we should inspect the channel types, and if there are no meg channels, we should plot the EEG, if we have meg only, then plot meg, then if we have both, probably the default should be meg but that can be discussed.

this should also be considered for other plotting functions that uses the `ch_type` argument.

### LINKED ISSUE #2113: channel position in plot_topomap does not always match head drawing
In plot_topomap, channel positions are rescaled to match head drawing. 

The rescaling is done here :
https://github.com/mne-tools/mne-python/blob/master/mne/viz/topomap.py#L240-L245
for each axis, it takes the min and the range, and stretch the coordinates so that the range cover the entire head.

This is fine when the montage is dense and "uniform", for example this one : 
image

but if you want a montage with channels only on the left hemisphere, this is what you get :
image

I have no perfect fix in mind. a cheap solution would be to add a option to disable this scaling.

### PR DESCRIPTION
**Pull request by Eric89GXL**
Closes #2113.
Closes #2027.

Example code:
```
import os.path as op
import mne
from mne.datasets.testing import data_path
evoked_fname = op.join(data_path(), 'MEG', 'sample', 'sample_audvis-ave.fif')
evoked = mne.read_evokeds(evoked_fname)[0]
evoked.pick_types(meg=False, eeg=True)
evoked.plot_topomap()
```
On current `master` this throws an error. This is what we now get on this PR:

figure_1

If change the last lines to select just the left EEG channels, with `ch_type='eeg'`:
```
left_ch = [ch['ch_name'] for ch in evoked.info['chs'] if ch['loc'][0] < 0]
evoked.pick_channels(left_ch)
evoked.plot_topomap()
```

On current master we get the left channels distributed all over the head, which is misleading:

figure_2

~~On this PR we get them staying on the left at least:~~

figure_3

~~Not sure if we also want to do the same thing with the y coords, since those could also errantly get distributed. Maybe this should also be a controllable option. Thoughts?~~

### CODE REVIEW

File: examples/datasets/plot_spm_faces_dataset.py [hunk 1]
```diff
@@ -97,8 +98,12 @@
...
 
 # Make source space
-src = mne.setup_source_space('spm', spacing='oct6', subjects_dir=subjects_dir,
-                             overwrite=True)
+src_fname = data_path + '/subjects/spm/bem/spm-oct-6-src.fif'
+if not os.path.isfile(src_fname):
+    src = mne.setup_source_space('spm', src_fname, spacing='oct6',
+                                 subjects_dir=subjects_dir, overwrite=True)
+else:
+    src = mne.read_source_spaces(src_fname)
```
Comment (Eric89GXL) [inline examples/datasets/plot_spm_faces_dataset.py]: Also snuck in this few-line edit that speeds up our `plot_spm_faces_dataset.py` example

File: mne/viz/tests/test_topomap.py [hunk 2]
```diff
@@ -55,91 +55,100 @@ def test_plot_topomap():
...
     # evoked
-    warnings.simplefilter('always', UserWarning)
+    warnings.simplefilter('always')
     res = 16
+    evoked = read_evokeds(evoked_fname, 'Left Auditory',
+                          baseline=(None, 0))
+    ev_bad = evoked.pick_types(meg=False, eeg=True, copy=True)
+    ev_bad.pick_channels(ev_bad.ch_names[:2])
+    ev_bad.plot_topomap()  # auto, should plot EEG
+    assert_raises(ValueError, ev_bad.plot_topomap, ch_type='mag')
```
Comment (Eric89GXL) [inline mne/viz/tests/test_topomap.py]: These four lines are the only new ones in this test, the rest were un-indenting a bunch of lines that really shouldn't have been indented by a `with warnings.catch_warnings` statement.

File: mne/viz/topomap.py [hunk 3]
```diff
@@ -1285,23 +1290,20 @@ def plot_psds_topomap(
...
         'Reds'.
-    agg_fun : callable
-        The function used to aggregate over frequencies.
-        Defaults to np.sum. if normalize is True, else np.mean.
     dB : bool
         If True, transform data to decibels (with ``10 * np.log10(data)``)
         following the application of `agg_fun`. Only valid if normalize is
         False.
-    n_jobs : int
-        Number of jobs to run in parallel.
```
Comment (Eric89GXL) [inline mne/viz/topomap.py]: ...and these docstrings were also incorrect

File: examples/datasets/plot_spm_faces_dataset.py (hunk 1 above)
Comment (dengemann) [inline examples/datasets/plot_spm_faces_dataset.py]: Thanks! Maybe stick with import os.path as op

File: mne/channels/channels.py [hunk 4]
```diff
@@ -69,17 +69,29 @@ def _contains_ch_type(info, ch_type):
         raise ValueError('`ch_type` is of class {actual_class}. It must be '
                          '`str`'.format(actual_class=type(ch_type)))
 
-    valid_channel_types = ('grad mag eeg stim eog emg ecg ref_meg resp '
-                           'exci ias syst seeg misc').split()
+    valid_channel_types = ('grad mag planar1 planar2 eeg stim eog emg ecg '
+                           'ref_meg resp exci ias syst seeg misc').split()
```
Comment (dengemann) [inline mne/channels/channels.py]: don't be so lazy :)

File: mne/viz/tests/test_topomap.py (hunk 2 above)
Comment (dengemann) [inline mne/viz/tests/test_topomap.py]: ok looks good

File: mne/viz/topomap.py [hunk 5]
```diff
@@ -237,11 +236,11 @@ def _check_outlines(pos, outlines, head_scale=0.85):
...
         x, y = pos[:, :2].T
-        x_range = np.abs(x.max() - x.min())
-        y_range = np.abs(y.max() - y.min())
+        x_range = max(x.max(), -x.min()) * 2.
+        y_range = y.max() - y.min()
 
         # shift and scale the electrode positions
-        pos[:, 0] = head_scale * ((pos[:, 0] - x.min()) / x_range - 0.5)
+        pos[:, 0] = head_scale * ((pos[:, 0]) / x_range)
         pos[:, 1] = head_scale * ((pos[:, 1] - y.min()) / y_range - 0.5)
```
Comment (dengemann) [inline mne/viz/topomap.py]: how is this stuff related?

Comment (dengemann) [inline mne/viz/topomap.py]: #compulsiveCleanup ?

File: mne/viz/topomap.py [hunk 6]
```diff
@@ -805,6 +808,8 @@ def plot_tfr_topomap(tfr, tmin=None, tmax=None, fmin=None, fmax=None,
     fig : matplotlib.figure.Figure
         The figure containing the topography.
     """
+    from ..channels import _get_ch_type
+    ch_type = _get_ch_type(tfr.info, ch_type)
     if format is not None:
         cbar_fmt = format
         warnings.warn("The format parameter is deprecated and will be "
```
Comment (dengemann) [inline mne/viz/topomap.py]: another format btw. we should fix this at some point ... (separate PR)

File: mne/viz/topomap.py (hunk 5 above)
Comment (Eric89GXL) [inline mne/viz/topomap.py]: this is the part of the code that forces the center to be @ x=0 (#2113) instead of in the middle of the (subset of the) sensor layout

Comment (Eric89GXL) [inline mne/viz/topomap.py]: we might actually want to do the same thing with the `y`, I'm not sure though

File: mne/viz/topomap.py (hunk 6 above)
Comment (Eric89GXL) [inline mne/viz/topomap.py]: yeah it is already deprecated, we need to make a sweeping deprecation-cleaning PR after release

File: mne/channels/channels.py (hunk 4 above)
Comment (Eric89GXL) [inline mne/channels/channels.py]: hey I'm just following the previous coder :) but yeah I agree it should be more explicit

File: mne/channels/channels.py [hunk 7]
```diff
@@ -69,17 +69,29 @@ def _contains_ch_type(info, ch_type):
...
+def _get_ch_type(info, ch_type):
+    """Helper to choose a singel channel type (usually for plotting)"""
+    if ch_type is None:
+        for type_ in ['mag', 'grad', 'planar1', 'planar2', 'eeg']:
+            if _contains_ch_type(info, type_):
+                ch_type = type_
+                break
+        else:
+            raise RuntimeError('No plottable channel types found')
+    return ch_type
```
Comment (dengemann) [inline mne/channels/channels.py]: @Eric89GXL so this is the main thing, right?
Maybe extend the comment to introduce future readers into the logic:

if we have mags, we will plot them,  if we have grads only, ...

File: mne/viz/topomap.py (hunk 5 above)
Comment (dengemann) [inline mne/viz/topomap.py]: ... separat issue ?? ;-) No I don't beat you to do that. Just another reminder.

File: mne/channels/channels.py (hunk 4 above)
Comment (dengemann) [inline mne/channels/channels.py]: ... ah damned that was me ...

**Comment by agramfort** [comment]
travis not happy

somebody please review and test. don't wait for me

File: mne/channels/channels.py (hunk 7 above)
Comment (teonlamont) [inline mne/channels/channels.py]: singel -> single

Comment (dengemann) [inline mne/channels/channels.py]: @Eric89GXL is there actually a reason why to run this at the info level and not at the instance level?
The code could be rewritten as follows:

```Python

def _get_ch_type(inst, ch_type):
    """Helper to choose a singel channel type (usually for plotting)"""
    if ch_type is None:
        for type_ in ['mag', 'grad', 'planar1', 'planar2', 'eeg']:
            if type_ in inst:
                ch_type = type_
                break
        else:
            raise RuntimeError('No plottable channel types found')
    return ch_type

```

Comment (Eric89GXL) [inline mne/channels/channels.py]: We just have to make sure all the instances support it, I can try it. It might fix the Travis failure, which appears to be due to `ica.info` being `None` (can that happen?)

Comment (dengemann) [inline mne/channels/channels.py]: I think most isntances have the contains mixin.

>  It might fix the Travis failure, which appears to be due to ica.info being None (can that happen?)

ICA does not have an info if there is no fit.

Comment (Eric89GXL) [inline mne/channels/channels.py]: How are we supposed to know if channels are in the ICA object, then? I get this error now that I've changed it to use `ch_type in ica`:
```
  File "/home/larsoner/custombuilds/mne-python/mne/preprocessing/ica.py", line 1329, in plot_components
    image_interp=image_interp)
  File "/home/larsoner/custombuilds/mne-python/mne/viz/topomap.py", line 638, in plot_ica_components
    ch_type = _get_ch_type(ica, ch_type)
  File "/home/larsoner/custombuilds/mne-python/mne/channels/channels.py", line 92, in _get_ch_type
    if type_ in inst:
  File "/home/larsoner/custombuilds/mne-python/mne/channels/channels.py", line 152, in __contains__
    has_ch_type = _contains_ch_type(self.info, ch_type)
  File "/home/larsoner/custombuilds/mne-python/mne/channels/channels.py", line 81, in _contains_ch_type
    return ch_type in [channel_type(info, ii) for ii in range(info['nchan'])]
TypeError: 'NoneType' object has no attribute '__getitem__'
```
The bug should be fixed at the `ContainsMixin` level probably.

Comment (dengemann) [inline mne/channels/channels.py]: Yes, if info is None, any channel in object should evaluate to False. No?

2015-05-19 19:24 GMT+02:00 Eric Larson <notifications@github.com>:

**Comment by Eric89GXL** [comment]
Comments addressed and the CIs are happy. Ready for review/merge from my end.

I made it so that both x=0 and y=0 are at the center of the head. I think this makes sense for cases like this:
```
import os.path as op
import mne
from mne.datasets.testing import data_path
evoked_fname = op.join(data_path(), 'MEG', 'sample', 'sample_audvis-ave.fif')
evoked = mne.read_evokeds(evoked_fname)[0]
evoked.pick_types(meg=False, eeg=True)
upper_left_ch = [ch['ch_name'] for ch in evoked.info['chs']
                 if ch['loc'][0] <= 0 and ch['loc'][1] >= 0]
evoked.pick_channels(upper_left_ch)
evoked.plot_topomap(ch_type='eeg')
```

**Comment by agramfort** [comment]
@alexandrebarachant can you give it a try?

File: mne/viz/topomap.py [hunk 8]
```diff
@@ -220,14 +220,26 @@ def plot_projs_topomap(projs, layout=None, cmap='RdBu_r', sensors=True,
     return fig
 
 
-def _check_outlines(pos, outlines, head_scale=0.85):
+def _check_outlines(pos, outlines, head_pos=None):
```
Comment (Eric89GXL) [inline mne/viz/topomap.py]: @alexandrebarachant the relevant changes start here

**Comment by agramfort** [comment]
Did not test but LGTM

**Comment by Eric89GXL** [comment]
Should we hold off on merging this until we release 0.9?

Speaking of which, @agramfort you want to do the "honors" of pushing to PyPi, etc.? Or maybe you have an idea of a good volunteer who could learn how to do it...?

**Comment by agramfort** [comment]
> Should we hold off on merging this until we release 0.9?

if @alexandrebarachant confirms it works for him, that tests pass and
the doc renders fine we can merge before release.

> Speaking of which, @agramfort you want to do the "honors" of pushing to PyPi, etc.? Or maybe you have an idea of a good volunteer who could learn how to do it...?

in my mind you would take care of it and have the honor to announce
the release on the mailing lists :)

but I can do it probably tomorrow.

**Comment by alexandrebarachant** [comment]
It works for me.
Just a quick thing, the default scalling seems to be a bit too small.

```
import numpy as np
from mne.channels import read_layout
from mne.viz import plot_topomap
layout = read_layout('biosemi')

plot_topomap(np.random.randn(64),layout.pos,vmin=-30,vmax=30)
```

image

Beside this, everything works fine :)

**Comment by agramfort** [comment]
how would you change the scaling?

**Comment by alexandrebarachant** [comment]
Seems that everyone was happy with the old one. I remember a value of 0.85 in the old code.

**Comment by agramfort** [comment]
let's wait for eric to wake up :)

**Comment by dengemann** [comment]
... please tell me onc you're happy. I will then have to check how this interacts with some extended options I've implemented for non-standard layouts like the EGI geodesic net. Also since I wrote the code in the first place maybe it would be a good idea to provide a final review.

**Comment by Eric89GXL** [comment]
Awake and fixed :) Feel free to try again @alexandrebarachant

**Comment by Eric89GXL** [comment]
@dengemann ready for review/merge from my end

File: examples/datasets/plot_spm_faces_dataset.py [hunk 9]
```diff
@@ -97,8 +98,12 @@
...
 
 # Make source space
-src = mne.setup_source_space('spm', spacing='oct6', subjects_dir=subjects_dir,
-                             overwrite=True)
+src_fname = data_path + '/subjects/spm/bem/spm-oct-6-src.fif'
+if not op.isfile(src_fname):
+    src = mne.setup_source_space('spm', src_fname, spacing='oct6',
+                                 subjects_dir=subjects_dir, overwrite=True)
+else:
+    src = mne.read_source_spaces(src_fname)
```
Comment (dengemann) [inline examples/datasets/plot_spm_faces_dataset.py]: .... such a common usecase... maybe worth a function in utils?

File: mne/evoked.py [hunk 10]
```diff
@@ -552,7 +559,8 @@ def plot_topomap(self, times=None, ch_type='mag', layout=None, vmin=None,
                                    mask_params=mask_params,
                                    outlines=outlines, contours=contours,
                                    image_interp=image_interp,
-                                   average=average, format=format)
+                                   average=average, head_pos=head_pos,
+                                   format=format)
```
Comment (dengemann) [inline mne/evoked.py]: `format`, really?

File: mne/viz/topomap.py [hunk 11]
```diff
@@ -220,14 +220,26 @@ def plot_projs_topomap(projs, layout=None, cmap='RdBu_r', sensors=True,
...
     """Check or create outlines for topoplot
     """
-    pos = np.asarray(pos)
+    pos = np.array(pos, float)[:, :2]  # ensure we have a copy
+    head_pos = {} if head_pos is None else head_pos
```
Comment (dengemann) [inline mne/viz/topomap.py]: dict()

File: mne/evoked.py (hunk 10 above)
Comment (dengemann) [inline mne/evoked.py]: ok deprecation cycle ...

File: mne/preprocessing/tests/test_ica.py [hunk 12]
```diff
@@ -203,6 +204,7 @@ def test_ica_core():
         with warnings.catch_warnings(record=True):
             ica.fit(raw, picks=pcks, start=start, stop=stop)
             repr(ica)  # to test repr
+        'mag' in ica  # should now work
```
Comment (dengemann) [inline mne/preprocessing/tests/test_ica.py]: assert_true?

File: mne/viz/topomap.py [hunk 13]
```diff
@@ -236,21 +248,21 @@ def _check_outlines(pos, outlines, head_scale=0.85):
...
 
         # shift and scale the electrode positions
-        pos[:, 0] = head_scale * ((pos[:, 0] - x.min()) / x_range - 0.5)
-        pos[:, 1] = head_scale * ((pos[:, 1] - y.min()) / y_range - 0.5)
+        if 'center' not in head_pos:
+            head_pos['center'] = 0.5 * (pos.max(axis=0) + pos.min(axis=0))
+        if 'scale' not in head_pos:
+            head_pos['scale'] = 0.85 / (pos.max(axis=0) - pos.min(axis=0))
+        pos -= head_pos['center']
+        pos *= head_pos['scale']
```
Comment (dengemann) [inline mne/viz/topomap.py]: where does 0.85 come from? this is new, right? Please add a parsable comment to the code.

File: mne/viz/topomap.py (hunk 3 above)
Comment (dengemann) [inline mne/viz/topomap.py]: thanks!

**Comment by dengemann** [comment]
Works fine down here with particular use cases. Besides my tiny comments good to go! Thanks @alexandrebarachant and @Eric89GXL !

File: mne/preprocessing/tests/test_ica.py (hunk 12 above)
Comment (Eric89GXL) [inline mne/preprocessing/tests/test_ica.py]: in this context the return type actually doesn't matter, it just matters that it succeeds -- but I can add the correct answer to be extra careful

File: examples/datasets/plot_spm_faces_dataset.py (hunk 9 above)
Comment (Eric89GXL) [inline examples/datasets/plot_spm_faces_dataset.py]: maybe eventually, but let's do it in another PR if someone feels strongly about it

File: mne/viz/topomap.py (hunk 13 above)
Comment (Eric89GXL) [inline mne/viz/topomap.py]: it comes from @agramfort according to `git blame` :) but that could just be due to refactoring or something. I can add a comment though

Comment (dengemann) [inline mne/viz/topomap.py]: What does it mean, why, where? :)

2015-05-21 3:13 GMT+02:00 Eric Larson <notifications@github.com>:

**Comment by Eric89GXL** [comment]
Comments addressed, thanks for the quick review @dengemann

File: mne/viz/topomap.py [hunk 14]
```diff
@@ -236,21 +248,23 @@ def _check_outlines(pos, outlines, head_scale=0.85):
...
-        y_range = np.abs(y.max() - y.min())
 
         # shift and scale the electrode positions
-        pos[:, 0] = head_scale * ((pos[:, 0] - x.min()) / x_range - 0.5)
-        pos[:, 1] = head_scale * ((pos[:, 1] - y.min()) / y_range - 0.5)
+        if 'center' not in head_pos:
+            head_pos['center'] = 0.5 * (pos.max(axis=0) + pos.min(axis=0))
+        if 'scale' not in head_pos:
+            # The default is to make the points occupy a slightly smaller
+            # proportion (0.85) of the total width and height
```
Comment (dengemann) [inline mne/viz/topomap.py]: I rember now that I put it there for some reason. Maybe extend the note saying that this is empirically determined and just works.

**Comment by dengemann** [comment]
Besides my last comment +1 for merge.

**Comment by Eric89GXL** [comment]
done

**Comment by dengemann** [comment]
Thanks a lot!!!
